### ISSUE TITLE
2019.01

### ISSUE THREAD

**Issue by mosra** [mosra/magnum#302]
I took a much-needed week-long break, which means 2018.12 is not humanly possible anymore and so the release will overflow to 2019. Happy new year, btw.

(Boring) things to do, in order:

- [x] ~~Make "shader compatibility" APIs and sources semi-public so these can be used from dependent libs and projects (TextureTools, imgui, Ui, <kbd>redacted</kbd> and probably more)~~ postponed
- [x] ImGui
    - [x] Merge imgui integration (https://github.com/mosra/magnum-integration/pull/33)
    - [x] ~~Port to WebGL 1 / ES2, enable that on CIs (depends on the public shader compat APIs)~~ postponed
    - [x] Merge imgui example (https://github.com/mosra/magnum-examples/pull/51)
    - [x] Make a WebGL demo for the website: https://magnum.graphics/showcase/imgui/
    - [x] Remove the links from community features, add all the people to website credits
    - [x] Update link to magnum integration in imgui's README, share the DPI-awareness workflow
    - [x] Update to ImGui 1.67, ~~enable mingw build~~ mosra/magnum-integration@01dc2b3f3295f21afa0297a7c52efc995c7d8b71, mingw already enabled
- [x] Everything else that's in the 2019.01 milestone and wasn't postponed (is anything else there?)
    - [x] Not anything significant from #293, as I want to have all these together to be able to measure the total improvement
- [x] Fix the Doxygen dumpster fire (the stable release is *again* anything but stable)
    - [x] Show `#include` for all classes, functions etc. (mosra/m.css#71)
- [x] ~~Make the Object Picking example work on Firefox (the `glReadPixels()` format fiasco again, https://github.com/mosra/magnum-website/issues/8)~~ impossible, hitting several bugs, postponed
- [x] Update toolchains submodule with Android / Emscripten on Windows fixes 
- [x] Document the "lack of model/view terminology" and explanation of inverse `lookAt()` in the transform docs, update the docs for #287 as well -- dba0928a145d1320095f3cab5b26cd42385c3d0e and 8240b8a60af17fcd209042c525885fa29b3e1219
- [x] ~~Format support tables (list of all formats with plugins that provide support)~~ i wanted them to look nice, but doxygen had a different opinion. Postponed.
- [x] Update bootstrap repos with dropped CMake 2.8 and GCC 4.7 support and updated toolchains submodule
- [x] Update copyright years (after all PRs are merged)
- [x] archlinux package update
- [x] homebrew package update -- mosra/homebrew-magnum@ded8123ac7b5a4467f1f84fa5898a2a6ac1dfb2a
- [x] vcpkg package update, dependent packages should depend only on `magnum[core]`, not the full thing (e.g. [magnum-integration](https://github.com/Microsoft/vcpkg/blob/master/ports/magnum-integration/CONTROL)) (@Squareys, thanks! :pray: ) -- https://github.com/microsoft/vcpkg/pull/5312
- [x] Hunter Corrade package update (https://github.com/ruslo/hunter/pull/1730)
- [x] Hunter Magnum package once everything is tagged (https://github.com/ruslo/hunter/pull/1731)
- [x] Rename `ScopedExit` to `ScopeGuard` mosra/corrade@f80dbfdeee4ff53dccf11ce8ede8fb915a7badfd

Nice to have:

- [x] ~~Crosscompilation support in Corrade Hunter package (details in #298)~~ postponed
- [x] Conan packages (#304) -- done for Corrade, Magnum package postponed
- [x] MSYS packages (mosra/corrade#56, mosra/magnum#307, mosra/magnum-plugins#50)
- [x] ~~SwiftShader-powered test on Travis (put a Linux `*.so` from Chrome binary on `ci.magnum.graphics`, use that), would bump test coverage from 52% to over 80%~~ postponed
- [x] ~~Make `GL::defaultFramebuffer` constexpr to avoid static constructors and linker issues~~ impossible, we need it mutable
- [x] ~~Fix Quaternion (s)lerp to not use atan (details in the reddit comment for the quaternion post)~~ i have the fix implemented already, but it needs more testing (especially for the error behavior), so [postponing](https://github.com/mosra/magnum/projects/5#card-17179459)
- [x] Initial work for the [magnum-singles](https://github.com/mosra/magnum-singles) repository
    - [x] Publish the single-header-generator tool ([docs](https://doc.magnum.graphics/corrade/acme.html))
    - [x] Write the README, reference it from Magnum readme
    - [x] ~~Add ArrayView library there~~ postponed
    - [x] Add ScopeGuard there
- [x] Make `ArrayView` slicing APIs `constexpr` (use the new assert) mosra/corrade@1158b099c5afc4326704cafa46835118a893660f
- [x] ~~STL compat for `ArrayView` (construct from vectors/arrays)~~ postponed, this needs quite a lot of work and testing to get right
- [x] ~~Make `Animation::TrackView` constexpr (needs constexpr type-erasing for `StridedArrayView`)~~ not possible due to function pointer casts, postponed
- [x] ~~File callbacks for `AbstractFont`~~ postponed

Fun (but not extremely important) things to do:

- [x] ~~Draft the ShaderTools library, add AbstractShaderConverter plugin iface~~ postponed
- [x] Conversions for basic Vulkan vector/rectangle/... types (Vulkan marketing!) -- 3281a277ccaec601fe016fe20886136ed169a136
- [x] `anyEvent()` in Sdl2App for easier handling of additional events -- 20a55b16aabf9d16af9a439b74f1260c15619c40
- [x] ~~`inotify` implementation for `Utility::FileWatcher` (and a blocking API for it)~~ postponed
- [x] ~~Windows implementation for `Utility::FileWatcher`~~ postponed
- [x] ~~Dynamic plugin support for Android 23+~~ postponed
- [x] ~~wrap up `FFmpegAudioImporter`, add image version of it (needs mainly documenting all the supported formats)~~ postponed
- [x] ~~Tweakable support for Android and Emscripten (need some way to supply changed files and notify the app ... (web)sockets?)~~ postponed
- [x] ~~Hot-reload support for Android (need some way to call `system()` on the host... sockets?)~~ postponed
- [x] ~~*Finally* integrate the Push The Box example~~ postponed

Use the opportunity to Break Stuff™ so I can more easily integrate stuff later (also fun):

- [x] ~~API-agnostic base for `Text::GlyphCache` (#143)~~ postponed
- [x] Design `Containers::UniquePtr` with opt-in (currently implicit) STL compatibility and convert all `Trade::AbstractImporter` APIs to it (could also help with #293 later)
    - [x] Move all APIs to use it
- [x] ~~Redesign `Trade::AbstractImageConverter` to make it possible to convert compressed images to compressed images and all other combinations~~ postponed
- [x] ~~Redesign `Trade::MeshData` to get rid of vectors, make it possible to support interleaved data similarly to `AnimationData` and make it more extensible for additional attributes~~ in progress, but a MeshTools rework needs to happen in parallel and that's quite a bit of work

**Comment by mosra** [comment]
@Squareys added a task for you: https://github.com/Microsoft/vcpkg/blob/master/ports/magnum-integration/CONTROL (and probably all others) should say just `depends: magnum[core]` instead of `depends: magnum`

(Thanks in advance! :pray: )

**Comment by mosra** [comment]
@pthom oh, thank you! I'm afraid the release tagging will happen more near to the end of the month, since there's still a lot of stuff I have to do until then.

The other Hunter-related task was crosscompilation support for Corrade (based on the hints `@ruslo` mentioned in #298), and that would be best to have done *before* the tagging happen so the released version can make use of it. But unless you explicitly need that yourself I don't want to force you into that task.

**Comment by pthom** [comment]
@mosra : ok. I do not need it at the moment. Ping me if you need help :-)

**Comment by mosra** [comment]
Progress so far:

- ImGui integration and the example is merged, web version available at https://magnum.graphics/showcase/imgui/. I'll postpone WebGL 1 / ES2 port for later, as it's not extremely critical.
- Copyright years updated :fireworks: 
- Doxygen post-release fixes submitted, issues opened (still have to use a fork, tho), the docs now tell what `#include` statements to use for all free symbols
- `Sdl2Application::anyEvent()` together with SDL_Event access done in 20a55b16aabf9d16af9a439b74f1260c15619c40
- `Containers::Pointer` and `Reference` implementations done ([article on the blog](https://blog.magnum.graphics/backstage/lightweight-stl-compatible-unique-pointer/)), now moving all APIs to them
- Initial work for the [magnum-singles](https://github.com/mosra/magnum-singles) repository done, generator tool published ([docs](https://doc.magnum.graphics/corrade/acme.html))

**Comment by mosra** [comment]
More progress:

- ImGui update to 1.67 done in mosra/magnum-integration@01dc2b3f3295f21afa0297a7c52efc995c7d8b71, one minor breaking change in a test
- `ScopedExit` -> `ScopeGuard` done in mosra/corrade@f80dbfdeee4ff53dccf11ce8ede8fb915a7badfd
- MSYS package PRs merged
- Constexpr ArrayView slicing done in mosra/corrade@1158b099c5afc4326704cafa46835118a893660f
- All APIs, examples etc. are moved to use the new Pointer and Reference APIs (CIs still running tho), except a few cases where it needs more work and thus was postponed
- [magnum-singles](https://github.com/mosra/magnum-singles) README etc. finished, `ScopeGuard` added there

Some larger work that needs some time to get right is postponed after the release.

**Comment by mosra** [comment]
@pthom The `v2019.01` tag is pushed now, so I guess ready for Hunter packaging. Sorry for this taking so long and thanks in advance! :+1:

**Comment by pthom** [comment]
@mosra: here is the PR for corrade v2019.01 : https://github.com/ruslo/hunter/pull/1730

**Comment by pthom** [comment]
@mosra : 
* PR for adding a new hunter package to the CI on ingenue : https://github.com/ingenue/hunter/pull/333
* PR for the new package itself on hunter: https://github.com/ruslo/hunter/pull/1731

**Comment by mosra** [comment]
@pthom thanks for all the work on the Hunter package :) It's now [in the docs](https://doc.magnum.graphics/magnum/building.html#building-packages-hunter) as well.

**Comment by mosra** [comment]
Arch packages (or at least the two main ones) [are updated now as well](https://www.archlinux.org/packages/?sort=&q=magnum&maintainer=&flagged=), so I think I can finally close this.

For the record, the release announcement is here: https://blog.magnum.graphics/announcements/2019.01/
